File: backend/routes/adminAccounts.js
/**
 * Admin Accounts Management Routes
 * Handles CRUD operations for admin accounts
 */

const express = require('express');
const admin = require('firebase-admin');
const AdminService = require('../services/adminService');

const router = express.Router();
const auth = admin.auth();

/**
 * Middleware to require superadmin role
 */
const requireSuperAdmin = async (req, res, next) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '');
    if (!token) {
      return res.status(401).json({ error: 'No authorization token' });
    }

    const decodedToken = await auth.verifyIdToken(token);
    const role = decodedToken.role || decodedToken.customClaims?.role;

    if (role !== 'superadmin') {
      return res.status(403).json({ error: 'Superadmin access required' });
    }

    req.adminUid = decodedToken.uid;
    next();
  } catch (error) {
    res.status(401).json({ error: error.message });
  }
};

/**
 * POST /api/admin-accounts/create
 * Create a new admin account
 * 
 * Request body:
 * {
 *   "email": "admin@wattwise.com",
 *   "displayName": "John Doe",
 *   "role": "admin|security|support|analyst"
 * }
 */
router.post('/create', requireSuperAdmin, async (req, res) => {
  try {
    const { email, displayName, role } = req.body;

    if (!email || !role) {
      return res.status(400).json({ error: 'Email and role are required' });
    }

    // Create Firebase Auth user
    const userRecord = await auth.createUser({
      email,
      password: Math.random().toString(36).slice(-12), // temporary password
      displayName: displayName || email,
    });

    // Set custom claims for role
    await auth.setCustomUserClaims(userRecord.uid, { role });

    // Create Firestore document
    const db = admin.firestore();
    await db.collection('roleBasedAccounts').doc(userRecord.uid).set({
      uid: userRecord.uid,
      email,
      displayName: displayName || email,
      role,
      status: 'active',
      createdAt: new Date(),
      updatedAt: new Date(),
      synced: true,
    });

    res.status(201).json({
      uid: userRecord.uid,
      email: userRecord.email,
      displayName: userRecord.displayName,
      role,
      message: 'Account created successfully',
    });
  } catch (error) {
    console.error('Create account error:', error);
    if (error.code === 'auth/email-already-exists') {
      res.status(400).json({ error: 'Email already exists' });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

/**
 * PUT /api/admin-accounts/:userId/update-role
 * Update the role of an account
 */
router.put('/:userId/update-role', requireSuperAdmin, async (req, res) => {
  try {
    const { role } = req.body;
    const userId = req.params.userId;

    if (!role) {
      return res.status(400).json({ error: 'Role is required' });
    }

    // Update custom claims
    await auth.setCustomUserClaims(userId, { role });

    // Update Firestore document
    const db = admin.firestore();
    await db.collection('roleBasedAccounts').doc(userId).update({
      role,
      updatedAt: new Date(),
    });

    res.status(200).json({
      message: 'Role updated successfully',
      userId,
      role,
    });
  } catch (error) {
    console.error('Update role error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/admin-accounts/:userId/disable
 * Disable an account
 */
router.post('/:userId/disable', requireSuperAdmin, async (req, res) => {
  try {
    const userId = req.params.userId;

    // Disable Firebase Auth user
    await auth.updateUser(userId, { disabled: true });

    // Update Firestore document
    const db = admin.firestore();
    await db.collection('roleBasedAccounts').doc(userId).update({
      status: 'disabled',
      updatedAt: new Date(),
    });

    res.status(200).json({
      message: 'Account disabled successfully',
      userId,
    });
  } catch (error) {
    console.error('Disable account error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * PUT /api/admin-accounts/:userId/update-display-name
 * Update display name
 */
router.put('/:userId/update-display-name', requireSuperAdmin, async (req, res) => {
  try {
    const { displayName } = req.body;
    const userId = req.params.userId;

    if (!displayName) {
      return res.status(400).json({ error: 'Display name is required' });
    }

    // Update Firebase Auth user
    await auth.updateUser(userId, { displayName });

    // Update Firestore document
    const db = admin.firestore();
    await db.collection('roleBasedAccounts').doc(userId).update({
      displayName,
      updatedAt: new Date(),
    });

    res.status(200).json({
      message: 'Display name updated successfully',
      userId,
      displayName,
    });
  } catch (error) {
    console.error('Update display name error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/admin-accounts/:userId/reset-password
 * Set a new password for the account
 */
router.post('/:userId/reset-password', requireSuperAdmin, async (req, res) => {
  try {
    const userId = req.params.userId;

    const { password } = req.body;

    if (!password || password.trim().length < 8) {
      return res.status(400).json({ error: 'Password must be at least 8 characters long' });
    }

    // Update user's password
    await auth.updateUser(userId, { password: password.trim() });

    res.status(200).json({
      message: 'Password updated successfully',
      userId,
    });
  } catch (error) {
    console.error('Reset password error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * DELETE /api/admin-accounts/:userId
 * Delete an admin account (superadmin only)
 */
router.delete('/:userId', requireSuperAdmin, async (req, res) => {
  try {
    const userId = req.params.userId;

    // Prevent deleting the superadmin that's doing the deletion
    const userRecord = await auth.getUser(userId);
    if (userRecord.customClaims?.role === 'superadmin' && userId === req.adminUid) {
      return res.status(400).json({ error: 'Cannot delete your own superadmin account' });
    }

    // Delete Firebase Auth user
    await auth.deleteUser(userId);

    // Delete Firestore document
    const db = admin.firestore();
    await db.collection('roleBasedAccounts').doc(userId).delete();

    res.status(200).json({
      message: 'Account deleted successfully',
      userId,
    });
  } catch (error) {
    console.error('Delete account error:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/admin/accounts
 * Create a new admin account
 * 
 * Request body:
 * {
 *   "email": "admin@wattwise.com",
 *   "displayName": "John Doe",
 *   "role": "admin|security|support|analyst"
 * }
 */
router.post('/', requireSuperAdmin, async (req, res) => {
  try {
    const { email, displayName, role } = req.body;

    if (!email || !displayName || !role) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    const result = await AdminService.createAdminAccount(
      email,
      displayName,
      role,
      req.adminUid
    );

    res.status(201).json(result);
  } catch (error) {
    if (error.message.includes('Invalid role')) {
      res.status(400).json({ error: error.message });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

/**
 * GET /api/admin/accounts
 * List all admin accounts
 * Query params:
 * - role: Filter by role (optional)
 */
router.get('/', requireSuperAdmin, async (req, res) => {
  try {
    // Temporarily disabled - use roleBasedAccounts collection directly
    res.status(200).json({ admins: [], total: 0 });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/admin/accounts/:adminId
 * Get admin account by UID
 */
router.get('/:adminId', requireSuperAdmin, async (req, res) => {
  try {
    // Temporarily disabled
    res.status(404).json({ error: 'Admin account not found' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * PUT /api/admin/accounts/:adminId
 * Update admin account
 * 
 * Request body:
 * {
 *   "displayName": "Updated Name",
 *   "role": "new_role",
 *   "isActive": true
 * }
 */
router.put('/:adminId', requireSuperAdmin, async (req, res) => {
  try {
    // Temporarily disabled
    res.status(500).json({ error: 'Endpoint temporarily disabled' });
  } catch (error) {
    if (error.message.includes('Invalid role')) {
      res.status(400).json({ error: error.message });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

/**
 * POST /api/admin/accounts/:adminId/deactivate
 * Deactivate an admin account
 */
router.post('/:adminId/deactivate', requireSuperAdmin, async (req, res) => {
  try {
    // Temporarily disabled
    res.status(500).json({ error: 'Endpoint temporarily disabled' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/admin/accounts/:adminId/reactivate
 * Reactivate an admin account
 */
router.post('/:adminId/reactivate', requireSuperAdmin, async (req, res) => {
  try {
    // Temporarily disabled
    res.status(500).json({ error: 'Endpoint temporarily disabled' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * DELETE /api/admin/accounts/:adminId
 * Permanently delete an admin account
 */
router.delete('/:adminId', requireSuperAdmin, async (req, res) => {
  try {
    // Temporarily disabled
    res.status(500).json({ error: 'Endpoint temporarily disabled' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: backend/routes/adminAccounts.py
"""
Admin Accounts Management Routes
Handles creation and management of admin/role-based accounts
"""

from flask import Blueprint, request, jsonify
import firebase_admin
from firebase_admin import auth, firestore
from datetime import datetime

admin_accounts_bp = Blueprint('admin_accounts', __name__)
db = firestore.client()

# Middleware to check superadmin role
def require_superadmin(f):
    from functools import wraps
    @wraps(f)
    def decorated_function(*args, **kwargs):
        token = request.headers.get('Authorization', '').replace('Bearer ', '')
        
        if not token:
            return jsonify({"error": "No authorization token"}), 401
        
        try:
            decoded_token = auth.verify_id_token(token)
            claims = decoded_token.get('custom_claims', {})
            
            if claims.get('role') != 'superadmin':
                return jsonify({"error": "Superadmin access required"}), 403
            
            request.admin_uid = decoded_token['uid']
            return f(*args, **kwargs)
        except Exception as e:
            return jsonify({"error": str(e)}), 401
    
    return decorated_function


@admin_accounts_bp.route('/create', methods=['POST'])
@require_superadmin
def create_admin_account():
    """
    Create a new admin/role-based account
    
    Request body:
    {
        "email": "admin@company.com",
        "role": "superadmin|admin|security|support|analyst|operator",
        "displayName": "John Doe (optional)",
        "password": "temporaryPassword123 (optional)"
    }
    """
    try:
        data = request.get_json()
        email = data.get('email')
        role = data.get('role')
        display_name = data.get('displayName', '')
        password = data.get('password', 'TempPass123!@#')  # Default temporary password
        
        # Validate required fields
        if not email or not role:
            return jsonify({"error": "email and role are required"}), 400
        
        # Validate role
        valid_roles = ['superadmin', 'admin', 'security', 'support', 'analyst', 'operator', 'user']
        if role not in valid_roles:
            return jsonify({"error": f"Invalid role. Must be one of: {', '.join(valid_roles)}"}), 400
        
        # Check if user already exists
        try:
            existing_user = auth.get_user_by_email(email)
            return jsonify({"error": f"User with email {email} already exists"}), 409
        except auth.UserNotFoundError:
            pass  # User doesn't exist, which is what we want
        
        # Create Firebase Auth user
        user = auth.create_user(
            email=email,
            password=password,
            display_name=display_name,
            email_verified=False
        )
        
        # Set custom claims for role
        custom_claims = {'role': role}
        auth.set_custom_user_claims(user.uid, custom_claims)
        
        # Create roleBasedAccounts Firestore document
        now = datetime.utcnow().isoformat()
        role_account = {
            'uid': user.uid,
            'email': email,
            'role': role,
            'displayName': display_name,
            'createdAt': now,
            'updatedAt': now,
            'createdBy': request.admin_uid,
            'status': 'active'
        }
        
        db.collection('roleBasedAccounts').document(user.uid).set(role_account)
        
        return jsonify({
            "success": True,
            "message": f"Account created for {email} with role: {role}",
            "user": {
                "uid": user.uid,
                "email": user.email,
                "displayName": user.display_name,
                "role": role,
                "createdAt": now
            }
        }), 201
    
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@admin_accounts_bp.route('/<user_id>/update-role', methods=['PUT'])
@require_superadmin
def update_account_role(user_id):
    """
    Update the role of an existing admin account
    
    Request body:
    {
        "role": "admin|security|support|analyst"
    }
    """
    try:
        data = request.get_json()
        new_role = data.get('role')
        
        if not new_role:
            return jsonify({"error": "role is required"}), 400
        
        valid_roles = ['superadmin', 'admin', 'security', 'support', 'analyst', 'operator', 'user']
        if new_role not in valid_roles:
            return jsonify({"error": f"Invalid role. Must be one of: {', '.join(valid_roles)}"}), 400
        
        # Update Firebase Auth custom claims
        user = auth.get_user(user_id)
        existing_claims = user.custom_claims or {}
        existing_claims['role'] = new_role
        auth.set_custom_user_claims(user_id, existing_claims)
        
        # Update Firestore document
        db.collection('roleBasedAccounts').document(user_id).update({
            'role': new_role,
            'updatedAt': datetime.utcnow().isoformat()
        })
        
        return jsonify({
            "success": True,
            "message": f"Role updated to {new_role}",
            "user": {
                "uid": user.uid,
                "email": user.email,
                "displayName": user.display_name,
                "role": new_role
            }
        }), 200
    
    except auth.UserNotFoundError:
        return jsonify({"error": f"User {user_id} not found"}), 404
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@admin_accounts_bp.route('/<user_id>/disable', methods=['POST'])
@require_superadmin
def disable_account(user_id):
    """
    Disable an admin account (soft delete)
    """
    try:
        # Disable Firebase Auth user
        auth.update_user(user_id, disabled=True)
        
        # Update Firestore status
        db.collection('roleBasedAccounts').document(user_id).update({
            'status': 'disabled',
            'updatedAt': datetime.utcnow().isoformat()
        })
        
        return jsonify({
            "success": True,
            "message": "Account disabled"
        }), 200
    
    except auth.UserNotFoundError:
        return jsonify({"error": f"User {user_id} not found"}), 404
    except Exception as e:
        return jsonify({"error": str(e)}), 500
